File: 2015/go/3/main-part-1.go
package main

import (
	"fmt"
	"os"
)

type Pos struct {
	x, y int
}

func main() {

	// get command line arguments
	if len(os.Args) < 2 {
		panic("Usage: go run main.go <input_file>")
	}

	var input = os.Args[1]

	// read input file
	file, err := os.Open(input)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var currentPos Pos = Pos{0, 0}
	var grid = make(map[int]map[int]int)
	grid[currentPos.x] = make(map[int]int)
	grid[currentPos.x][currentPos.y] = 1

	// parse input
	for {
		var line string
		_, err := fmt.Fscanln(file, &line)
		if err != nil {
			break
		}

		for _, c := range line {
			switch c {
			case '^':
				currentPos.y++
			case 'v':
				currentPos.y--
			case '>':
				currentPos.x++
			case '<':
				currentPos.x--
			default:
				panic("Invalid character")
			}

			if _, ok := grid[currentPos.x]; !ok {
				grid[currentPos.x] = make(map[int]int)
			}

			grid[currentPos.x][currentPos.y]++
		}
	}

	count := 0
	for _, row := range grid {
		count += len(row)
	}

	fmt.Println("Part 1:", count)
}

File: 2015/zig/3/main-part-1.zig
const std = @import("std");

const Position = struct {
    x: i32,
    y: i32,

    pub fn hash(self: Position) u64 {
        return std.mem.hash(self);
    }

    pub fn eq(self: @This(), other: @This()) bool {
        return self.x == other.x and self.y == other.y;
    }

    pub fn move(self: Position, c: u8) Position {
        switch (c) {
            'v' => return Position{ .x = self.x, .y = self.y - 1 },
            '^' => return Position{ .x = self.x, .y = self.y + 1 },
            '<' => return Position{ .x = self.x - 1, .y = self.y },
            '>' => return Position{ .x = self.x + 1, .y = self.y },
            else => return self,
        }
    }
};

pub fn hashPosition(items: []Position, value: Position) bool {
    for (items) |item| {
        if (item.eq(value)) {
            return true;
        }
    }
    return false;
}

pub fn main() !void {
    // get command line arguments
    const stdout = std.io.getStdOut().writer();
    const args = try std.process.argsAlloc(std.heap.page_allocator);
    defer std.process.argsFree(std.heap.page_allocator, args);

    if (args.len < 2) {
        try stdout.print("Usage: zig run main.zig <input_file>\n", .{});
        return;
    }

    const file_path = args[1];
    const file = try std.fs.cwd().openFile(file_path, .{});
    defer file.close();

    const file_contents = try file.reader().readAllAlloc(std.heap.page_allocator, 10000);

    var position: Position = Position{ .x = 0, .y = 0 };

    var visited = std.ArrayList(Position).init(std.heap.page_allocator);
    defer visited.deinit();

    try visited.append(position);

    for (file_contents) |c| {
        position = position.move(c);
        if (hashPosition(visited.items, position)) {
            continue;
        }
        try visited.append(position);
    }

    try stdout.print("Part 1: {}\n", .{visited.items.len});
}
